File: python/Gray Code.py
class Solution(object):
    def grayCode(self, n):
        if n == 0:
            return [0]
        if n == 1:
            return [0, 1]
        else:
            base = self.grayCode(n - 1)
            return [num for num in base] + [2 ** (n-1) + num for num in base[::-1]]

File: cpp/Gray Code.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        vector<int> result(1, 0);
        /** level-by-level expand the result **/
        /** add 1 to prefix of each int the previous layer ***/
        for (int i = 0; i < n; i++) {
            int cur = result.size();
            while (cur) {
                cur--;
                int temp = result[cur];
                temp += (1<<i);
                result.push_back(temp);
            }
        }
        return result;
    }
};
